examples/01-blog/Blog/Data/DataSource.php

```diff
@@ -1,202 +1,250 @@
 <?php
+
+declare(strict_types=1);
+
 namespace GraphQL\Examples\Blog\Data;
 
+use function array_filter;
+use function array_keys;
+use function array_map;
+use function array_search;
+use function array_slice;
+use function array_values;
+use function count;
+use function in_array;
+use function rand;
+
 /**
  * This is just a simple in-memory data holder for the sake of example.
  * Data layer for real app may use Doctrine or query the database directly (e.g. in CQRS style)
  */
 class DataSource
 {
-    private static $users = [];
-    private static $stories = [];
-    private static $storyLikes = [];
-    private static $comments = [];
-    private static $storyComments = [];
-    private static $commentReplies = [];
-    private static $storyMentions = [];
-
-    public static function init()
+    /** @var array<int, User> */
+    private static array $users = [];
+
+    /** @var array<int, Story> */
+    private static array $stories = [];
+
+    /** @var array<int, array<int, int>> */
+    private static array $storyLikes = [];
+
+    /** @var array<int, Comment> */
+    private static array $comments = [];
+
+    /** @var array<int, array<int, int>> */
+    private static array $storyComments = [];
+
+    /** @var array<int, array<int, int>> */
+    private static array $commentReplies = [];
+
+    /** @var array<int, array<int, User|Story>> */
+    private static array $storyMentions = [];
+
+    public static function init(): void
     {
         self::$users = [
-            '1' => new User([
-                'id' => '1',
+            1 => new User([
+                'id' => 1,
                 'email' => 'john@example.com',
                 'firstName' => 'John',
-                'lastName' => 'Doe'
+                'lastName' => 'Doe',
             ]),
-            '2' => new User([
-                'id' => '2',
+            2 => new User([
+                'id' => 2,
                 'email' => 'jane@example.com',
                 'firstName' => 'Jane',
-                'lastName' => 'Doe'
+                'lastName' => 'Doe',
             ]),
-            '3' => new User([
-                'id' => '3',
+            3 => new User([
+                'id' => 3,
```

Comment:
Having those IDs as strings were IMO better. The fact it's numeric is a coincidence.

Reply:
GraphQL generally seems to push towards this with the ID type. In practice, I found that databases are more performant when using numeric IDs and tend to use them pervasively.

In this example, there is no big difference. Would be fine as it was before, is also fine like this.

Reply:
> I found that databases are more performant when using numeric IDs and tend to use them pervasively.

side note, depends significantly on the database, no?

Reply:
Indices on numeric values are generally more efficient, subsequently `WHERE IN` or `JOIN` or faster.